shelley/chain-and-ledger/executable-spec/test/Test/Shelley/Spec/Ledger/SerializationProperties.hs

```diff
@@ -0,0 +1,326 @@
+import Test.QuickCheck
+  ( Arbitrary,
+    arbitrary,
+    genericShrink,
+    oneof,
+    shrink,
+  )
+import Test.QuickCheck.Hedgehog (hedgehog)
+import qualified Test.Shelley.Spec.Ledger.ConcreteCryptoTypes as Mock
+  ( Addr,
+    BHeader,
+    Block,
+    ConcreteCrypto,
+    DELEG,
+    DELEGS,
+    DELPL,
+    HashHeader,
+    LEDGER,
+    LEDGERS,
+    POOL,
+    PPUP,
+    Tx,
+    TxId,
+    TxOut,
+    UTXO,
+    UTXOW,
+  )
+import Test.Shelley.Spec.Ledger.Generator.Core (KeySpace (KeySpace_), NatNonce (..), geKeySpace, ksCoreNodes, mkBlock, mkOCert)
+import Test.Shelley.Spec.Ledger.Generator.Presets (genEnv)
+import Test.Shelley.Spec.Ledger.NonTraceProperties.Generator (genStateTx)
+import Test.Tasty.QuickCheck ((===), Gen, Property, choose, counterexample, elements)
+
+roundtrip ::
+  (Eq a, Show a) =>
+  (a -> Encoding) ->
+  (forall s. Decoder s a) ->
+  a ->
+  Property
+roundtrip enc dec = roundtrip' enc (const <$> dec)
+
+-- | Roundtrip property for values annotated with their serialized form
+--
+-- NOTE: Suppose @a@ consists of a pair of the unannotated value @a'@ and some
+-- 'Lazy.ByteString'. The roundtrip property will fail if that
+-- 'Lazy.ByteString' encoding is not equal to @enc a'@. One way in which this
+-- might happen is if the annotation is not canonical CBOR, but @enc@ does
+-- produce canonical CBOR.
+roundtrip' ::
+  (Eq a, Show a) =>
+  -- | @enc@
+  (a -> Encoding) ->
+  (forall s. Decoder s (Lazy.ByteString -> a)) ->
+  a ->
+  Property
+roundtrip' enc dec a = case deserialiseFromBytes dec bs of
+  Right (bs', a')
+    | Lazy.null bs' ->
+      a === a' bs
+    | otherwise ->
+      counterexample ("left-over bytes: " <> show bs') False
+  Left e ->
+    counterexample (show e) False
+  where
+    bs = toLazyByteString (enc a)
+
+genHash :: forall a c. Crypto c => Proxy c -> Gen (Hash c a)
+genHash proxy = mkDummyHash proxy <$> arbitrary
+
+mkDummyHash :: forall c a. Crypto c => Proxy c -> Int -> Hash c a
+mkDummyHash _ = coerce . hash @(HASH c)
+
+{-------------------------------------------------------------------------------
+  Serialization Properties
+-------------------------------------------------------------------------------}
+
+prop_roundtrip_Block :: Mock.Block -> Property
+prop_roundtrip_Block = roundtrip' toCBOR ((. Full) . runAnnotator <$> fromCBOR)
+
+prop_roundtrip_Header :: Mock.BHeader -> Property
+prop_roundtrip_Header = roundtrip' toCBOR ((. Full) . runAnnotator <$> fromCBOR)
+
+prop_roundtrip_HeaderHash :: Mock.HashHeader -> Property
+prop_roundtrip_HeaderHash = roundtrip toCBOR fromCBOR
+
+prop_roundtrip_Tx :: Mock.Tx -> Property
+prop_roundtrip_Tx = roundtrip' toCBOR ((. Full) . runAnnotator <$> fromCBOR)
+
+prop_roundtrip_TxId :: Mock.TxId -> Property
+prop_roundtrip_TxId = roundtrip toCBOR fromCBOR
+
+prop_roundtrip_ApplyTxErr :: [STS.PredicateFailure Mock.LEDGERS] -> Property
+prop_roundtrip_ApplyTxErr = roundtrip toCBOR fromCBOR
```

Comment:
Again, some of these names are just copy-pasted from consensus and should be updated.